### PR TITLE
Fix compiling with Trilinos+Sacado+Cuda

### LINKED ISSUE #14487: Compilation errors in Trilinos with Cuda support
I built Trilinos with Cuda (version 11.1.1) support and then built deal.ii using the following script
```
cmake  -D CMAKE_INSTALL_PREFIX=../installed-candi-cuda \
              -D CMAKE_CXX_COMPILER=$TRILINOS_DIR/bin/nvcc_wrapper \
              -D DEAL_II_CXX_FLAGS='-Werror -Wno-non-template-friend' \
              -D DEAL_II_EARLY_DEPRECATIONS=ON \
              -D DEAL_II_WITH_CUDA="ON" \
              -D DEAL_II_WITH_KOKKOS="ON" \
              -D KOKKOS_DIR=$TRILINOS_DIR \
              -D DEAL_II_WITH_MPI="ON" \
              -D DEAL_II_WITH_P4EST="ON" \
              -D DEAL_II_WITH_TRILINOS="ON" \
              -D DEAL_II_COMPONENT_EXAMPLES="OFF" \
	      ../dealii
```
but got the following errors
```

/home/jiaqi2/dealii/dealii/include/deal.II/base/tensor.h(1651): error: identifier "Sacado::Rad::ADvari<double> ::adc" is undefined in device code

/home/jiaqi2/software/gcc-kokkos/trilinos-release-13-2-0/include/Sacado_trad.hpp(468): error: identifier "Sacado::Rad::ADcontext<double> ::One" is undefined in device code

/home/jiaqi2/software/gcc-kokkos/trilinos-release-13-2-0/include/Sacado_trad.hpp(468): error: identifier "Sacado::Rad::Derp<double> ::LastDerp" is undefined in device code

/home/jiaqi2/software/gcc-kokkos/trilinos-release-13-2-0/include/Sacado_trad.hpp(468): error: identifier "Sacado::Rad::Derp<double> ::LastDerp" is undefined in device code

```

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #14487. I need the first commit for it to compile. Apparently, we can do without the code the second commit removes as well.

### CODE REVIEW

File: include/deal.II/differentiation/ad/ad_helpers.h [hunk 1]
```diff
@@ -2837,8 +2837,7 @@ namespace Differentiation
        * @note For taped AD numbers, this operation is only valid in recording mode.
        */
       template <typename ExtractorType>
-      typename internal::Extractor<dim,
-                                   ExtractorType>::template tensor_type<ad_type>
+      auto
       get_sensitive_variables(const ExtractorType &extractor) const;
```
Comment (drwells) [inline include/deal.II/differentiation/ad/ad_helpers.h:2841]: Can we use a more descriptive return type?

Comment (masterleinad) [reply include/deal.II/differentiation/ad/ad_helpers.h:2842]: I found an alternative that the compiler accepts.

**Comment by masterleinad** [comment]
Only `parallel_vector_22` and `parallel_vector_23` are failing and we know that that is already fixed on `master`.

**Review by drwells — APPROVED** [review]

**Comment by masterleinad** [comment]
Could I get another review for this?

**Comment by drwells** [comment]
We have enough GPU patches coming and going now that I think its wiser to just merge this now and handle further tweaks in a followup.

**Pull request merged**